File: c++/Data_Struct/LL.cpp
#include<iostream>

using namespace std;

void insert();
void deletion();
void create();
void Display();
void Clear();

struct Node{
    public:
        int data;
        Node *next;
    };
    
Node* head = NULL;
Node* tail = NULL;



int main(){
    while(1){
    int choice;
    cout<<"1. Insertion"<<endl;
    cout<<"2. Deletion"<<endl;
    cout<<"3. create"<<endl;
    cout<<"4. Display"<<endl;
    cout<<"5. Delete list"<<endl;

    cout<<"Enter the choice: ";
    cin>>choice;
    switch (choice)
    {
    case 1:
        insert();
        break;
    case 2:
        deletion();
        break;
    case 3:
        create();
        break;
    case 4:
        Display();
        break;
    default:
        Clear();
        exit(0);
    }
}
}


void create(){
    int n;
    cout<<"Enter the number of nodes: ";
    cin>>n;
    Node *temp;
    for(int i=0;i<n;i++){
        temp=new Node;
        cout<<"Enter the "<<i+1<<" element: ";
        cin>>temp->data;
        temp->next=NULL;
        if(head==NULL){
            head=temp;
            tail=temp;
        }
        else{
            tail->next=temp;
            tail=temp;
        }
       
    }
}

void insert(){
    Node *temp=new Node,*trav;
    int a,choice;
    trav=head;
    cout<<"Enter the element/location: ";
    cin>>a;
    cout<<"1. Insertion in front"<<endl<<"2. Insertion at the end"<<endl;
    if(choice==2){
        while (trav->data!=a)
        {
            trav=trav->next;
        }
        temp->next=trav->next;
        trav->next=temp;
        if(trav==tail){
            tail=temp;
        }
        if(trav==head){
            head=temp;
            temp->next=trav;
        }
        cout<<"enter the element: ";
        cin>>temp->data;
        }
    else{
        //insert before the element
        while (trav->next->data!=a)
        {
            trav=trav->next;
        }
        temp->next=trav->next;
        trav->next=temp;
        cout<<"enter the data: ";
        cin>>temp->data;

    }    
}

void deletion(){
    Node *temp,*tr;
    int a;
    tr=head;
    cout<<"Enter the element/location: ";
    cin>>a;
    while (tr->data!=a)
    {
        temp=tr;
        tr=tr->next;
    }
    temp->next=tr->next;
    delete tr;
}

void Display(){
    Node *trav;
    trav=head;
    while (trav!=NULL)
    {
        cout<<trav->data<<" "<<endl;
        trav=trav->next;
    }
}

void Clear(){
    Node *temp=new Node,*trav;
    trav=head;
    while (trav!=NULL)
    {
        temp=trav;
        trav=trav->next;
        delete temp;
    }
}

File: py/DS/LL.py


class Node:
    def __init__(self,x=None):
        
        self.data = x
        self.next = None
Head,Tail =None,None


def Create():
        global Head,Tail
        n = int(input("size"))
        while(n>0):
            temp = Node()
            if Head is None:
                Head = temp
                Tail = temp
            else:   
                Tail.next = temp
                Tail = temp
            temp.data = int(input("data"))
            n-=1
            
def Insert(loc,node:Node = Node(),pos = 2):
    """
    loc = location
    node = node to be inserted
    pos = 1 for before node
    pos = 2 for after node
    """
    global Head,Tail
    if(pos==1): # insert at begining
        
        tr = Head
        while (tr.data!=loc):
            tr = tr.next
        temp = tr.next
        tr.next = node
        node.next = temp
    else:
        
        tr = Head
        while (tr.data!=loc):
            tr = tr.next
        temp = tr.next
        tr.next = node
        node.next = temp
        if(tr==Tail):
            Tail = node

def Delete(loc,data):
        global Head,Tail
        temp = Head
        while(temp.data!=loc):
            temp = temp.next
        temp.next = temp.next.next
        if(temp==Tail):
            Tail = temp
        temp = temp.next
        if(temp.data==data):
            temp = temp.next
            Head = temp
        else:
            temp = temp.next
            while(temp.data!=data):
                temp = temp.next
            temp.next = temp.next.next
            if(temp.next==None):
                Tail = temp
                
            
def Print():
        global Head,Tail
        temp = Head
        while(temp!=Tail.next):
            print(f"{temp.data}",end="-> ")
            temp = temp.next
        print("\n")
        

Create()
x = int(input("data to be inserted"))
y =int(input("insert 1.before or 2.after"))
Insert(2,Node(x),y)


Print()
